### ISSUE TITLE
Useful functions missing in the String module?

### ISSUE THREAD

**Issue by eriksvedang** [carp-lang/Carp#94]
The String module is probably lacking a lot of useful functions commonly found in other standard libraries. Let's add them!

**Comment by hellerve** [comment]
I have a list of useful functions (this is a verbatim copy of what’s present in zepto):

- `(String.join strs delim)`: joins a list of strings with a delimiter
- `(String.split str delim)`: splits a string on a delimiter, returns a list of substrings
- `(String.in? str sub)`: checks whether a substring is within a string
- `(String.count str c)`: counts the number of occurrences of character c in a string
- `(String.starts-with? str sub)`: checks whether a string starts with a substring
- `(String.ends-with? str sub)`: checks whether a string ends with a substring
- `(String.trim str)`: trims whitespace from a string
- `(String.upper? str)`: checks whether string is upper case
- `(String.lower? str)`: checks whether string is lower case
- `(String.num? str)`: checks whether string is numerical
- `(String.alpha? str)`: checks whether string is alphabetical
- `(String.alphanum? str)`: checks whether string is alphanumerical

There are more useful functions, but that’s a whole lot and a good start, I’d say.

Cheers

**Comment by eriksvedang** [comment]
Great, yes this looks like a very decent start indeed.

**Comment by Kungsgeten** [comment]
Maybe [s.el](https://github.com/magnars/s.el) (the long lost Emacs string manipulation library) can be used as inspiration?

**Comment by hellerve** [comment]
Looks good! It also reminds me that we still have to wrap a regex library!

**Comment by bjorn3** [comment]
* string to int
* int to string

**Comment by hellerve** [comment]
Those are already provided as `Int.str` and `Int.from-string`! The floating point data types do not have `from-string` functions, but should probably have them...

**Comment by bjorn3** [comment]
Should have looked at the source

**Comment by hellerve** [comment]
Don’t worry! We have to document the standard modules sooner or later, to make it a bit easier for people to know what’s there.

**Comment by eriksvedang** [comment]
I feel like this is OK to close now?

**Comment by hellerve** [comment]
I haven’t ported everything from the long lost Emacs string library yet, nor do I intend to. SOme of the functions do not need to b included in the standard library, in my opinion.

That being said, I do think that a handful of functions from the library should be included, like `truncate` and `split-up-to`.

**Comment by eriksvedang** [comment]
OK, let us leave it open until then, then :)

**Comment by mkfifo** [comment]
I am going to do a little work on the Strings module, adding documentation, and adding some new functions which I think are important.

I will send a PR once I have a few commits ready.

**Comment by mkfifo** [comment]
Sorry for the noise, lots of rebasing to get things to line up nicely.

**Comment by eriksvedang** [comment]
I guess the String module will never be completely "done", so perhaps I can close this now? Or are there some functions that we really need, still? If so – let's enumerate them so we know what to shoot for.

**Comment by mkfifo** [comment]
@eriksvedang good point, I'm pretty happy with the current shape of String for now, I don't see anything huge missing, happy for this to be closed.
